hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/ExpressionIndexSupport.scala

```diff
@@ -92,6 +95,47 @@ class ExpressionIndexSupport(spark: SparkSession,
     }
   }
 
+  def prunePartitions(fileIndex: HoodieFileIndex,
+                      queryFilters: Seq[Expression],
+                      queryReferencedColumns: Seq[String]): Option[Set[String]] = {
+    lazy val expressionIndexPartitionOpt = getExpressionIndexPartitionAndLiterals(queryFilters)
+    if (isIndexAvailable && queryFilters.nonEmpty && expressionIndexPartitionOpt.nonEmpty) {
+      val (indexPartition, expressionIndexQuery, _) = expressionIndexPartitionOpt.get
+      val indexDefinition = metaClient.getIndexMetadata.get().getIndexDefinitions.get(indexPartition)
+      if (indexDefinition.getIndexType.equals(HoodieTableMetadataUtil.PARTITION_NAME_COLUMN_STATS)) {
+        val readInMemory = shouldReadInMemory(fileIndex, queryReferencedColumns, inMemoryProjectionThreshold)
+        val expressionIndexRecords = loadExpressionIndexPartitionStatRecords(indexDefinition, readInMemory)
+        loadTransposed(queryReferencedColumns, readInMemory, expressionIndexRecords, expressionIndexQuery) {
+          transposedPartitionStatsDF => {
+            val allPartitions = transposedPartitionStatsDF.select(HoodieMetadataPayload.COLUMN_STATS_FIELD_FILE_NAME)
+              .collect()
+              .map(_.getString(0))
+              .toSet
+            if (allPartitions.nonEmpty) {
+              // NOTE: [[translateIntoColumnStatsIndexFilterExpr]] has covered the case where the
+              //       column in a filter does not have the stats available, by making sure such a
+              //       filter does not prune any partition.
+              val indexSchema = transposedPartitionStatsDF.schema
+              val indexFilter = Seq(expressionIndexQuery).map(translateIntoColumnStatsIndexFilterExpr(_, indexSchema, isExpressionIndex = true)).reduce(And)
+              Some(transposedPartitionStatsDF.where(new Column(indexFilter))
+                .select(HoodieMetadataPayload.COLUMN_STATS_FIELD_FILE_NAME)
+                .collect()
```

Comment:
I guess this is a remnant of partitions stats index support - `collect` called on same `transposedPartitionStatsDF` repeatedly? Also, is this on driver? Should we cache first?

Reply:
Addressed for both EI and partition stats